packages/components/src/border-control/border-control-style-picker/component.tsx

```diff
@@ -0,0 +1,87 @@
+/**
+ * WordPress dependencies
+ */
+import { lineDashed, lineDotted, lineSolid } from '@wordpress/icons';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Button from '../../button';
+import { StyledLabel } from '../../base-control/styles/base-control-styles';
+import { View } from '../../view';
+import { Flex } from '../../flex';
+import { VisuallyHidden } from '../../visually-hidden';
+import { contextConnect, WordPressComponentProps } from '../../ui/context';
+import { useBorderControlStylePicker } from './hook';
+
+import type { LabelProps, StylePickerProps } from '../types';
+
+const BORDER_STYLES = [
+	{ label: __( 'Solid' ), icon: lineSolid, value: 'solid' },
+	{ label: __( 'Dashed' ), icon: lineDashed, value: 'dashed' },
+	{ label: __( 'Dotted' ), icon: lineDotted, value: 'dotted' },
+];
+
+const Label = ( props: LabelProps ) => {
+	const { label, hideLabelFromVision } = props;
+
+	if ( ! label ) {
+		return null;
+	}
+
+	return hideLabelFromVision ? (
+		<VisuallyHidden as="label">{ label }</VisuallyHidden>
+	) : (
+		<StyledLabel>{ label }</StyledLabel>
+	);
+};
+
+const BorderControlStylePicker = (
```

Comment:
> As a follow-up, we'll probably want to look into componentizing this pattern (related: [#36735](https://github.com/WordPress/gutenberg/issues/36735)).

Agreed. It's a little tricky to reconcile this pattern. The main problem being it can't really be a radio group (and therefore not a `ToggleGroupControl` ) as we need to be able to toggle off a selection. 

The idea of adding a "none" option so we can have that as a default doesn't in itself work for some real-world cases such as allowing values to be inherited from CSS or Global Styles in the block editor. 

We could add an `inherit` option as well to cover that but now we've nearly doubled the number of buttons in the control. The problem is also compounded by the fact that its difficult to have icons that represent `none` or `inherit`, particularly when a minus icon as suggested in #36735 is almost the same as the `solid` style's icon.

> I also noticed that a click on a color doesn't automatically select the `solid` option:

This was by design. 

I initially made the component force a selection however the resulting experience was very clunky. It was actually very frustrating to use in the real world case driving the creation of this component.

When it comes to the block editor we need to be able to inherit the border style from possibly theme provided CSS or Global Styles. Forcing a style selection would override that inheritance potentially against the user's wishes. It became more of a problem when you then tried to toggle off a border style selection and suddenly a new style was selected. 

For example, selecting a color forces solid style but the user wants it inherited from CSS, they try to toggle off the selection but it appears stuck. Another behaviour encountered was for the user to then select a different style, e.g. `dashed`, that matches the global style, then they realize they should just inherit it so it can be changed as needed globally, they try to toggle off the `dashed` selection and frustratingly the `solid` option was selected automatically due to an existing color selection. Only setting a style selection on the initial change of color then often just meant extra clicks to undo the automated style selection. 

The end result was to remove any automated selections and stick close to the behaviour of the existing border-style control from the block editor.

I'd suggest we could look at refining this in a follow-up. What do you think?

Reply:
> It's a little tricky to reconcile this pattern.

Very much so 😕 It almost feels like a "UI smell" that we have this much trouble expressing the UI in a standards-compliant way.

> This was by design.

Thanks, your explanation makes sense, and I think I understood the inheritance concept better after reviewing `BorderBoxControl` (#38876). I think what threw me off the most was how the outer ring on the indicator still shows as `solid` even though the `solid` style is not explicitly selected. Would it make sense to _not_ show that outer ring at all when there's no style selected?

![Outer ring of the color indicator is showing the solid style](https://user-images.githubusercontent.com/555336/158604997-4520820c-dda5-4beb-9e5f-b0317202c021.png)

Reply:
I've tweaked the indicator to only show the border when there has been a style selection. Let me know if you think we still need further refinements.

https://user-images.githubusercontent.com/60436221/158947836-3eeb770b-18a9-4d98-bff0-fe31e4a9ce87.mp4

Reply:
Awesome, this makes perfect sense now!